controller/api/public/grpc_server_test.go

```diff
@@ -186,6 +189,131 @@ spec:
 			}
 		}
 	})
+
+	t.Run("Fails for both resource and namespace set in request", func(t *testing.T) {
+		k8sAPI, err := k8s.NewFakeAPI("", make([]string, 0)...)
+		if err != nil {
+			t.Fatalf("NewFakeAPI returned an error: %s", err)
+		}
+
+		fakeGrpcServer := newGrpcServer(
+			&MockProm{Res: model.Vector{
+				&model.Sample{
+					Metric:    model.Metric{"pod": "emojivoto-meshed"},
+					Timestamp: 456,
+				},
+			}},
+			tap.NewTapClient(nil),
+			k8sAPI,
+			"linkerd",
+			[]string{},
+		)
+
+		k8sAPI.Sync()
+
+		_, err = fakeGrpcServer.ListPods(context.TODO(), &pb.ListPodsRequest{
+			Namespace: "test",
+			Selector: &pb.ResourceSelection{
+				Resource: &pb.Resource{
+					Type: pkgK8s.Pod,
+				},
+			},
+		})
+		expectedError := fmt.Errorf("cannot set both namespace and resource in the request. These are mutually exclusive")
+		if err.Error() != expectedError.Error() {
+			t.Fatalf("Expected error: %s, Got: %s", expectedError, err)
+		}
+	})
+
+	t.Run("Reads namespace from resource namespace field", func(t *testing.T) {
+		k8sAPI, err := k8s.NewFakeAPI("", make([]string, 0)...)
+		if err != nil {
+			t.Fatalf("NewFakeAPI returned an error: %s", err)
+		}
+
+		mockProm := MockProm{Res: model.Vector{
+			&model.Sample{
+				Metric:    model.Metric{"pod": "emojivoto-meshed"},
+				Timestamp: 456,
+			},
+		}}
+		fakeGrpcServer := newGrpcServer(
+			&mockProm,
+			tap.NewTapClient(nil),
+			k8sAPI,
+			"linkerd",
+			[]string{},
+		)
+
+		k8sAPI.Sync()
+
+		_, _ = fakeGrpcServer.ListPods(context.TODO(), &pb.ListPodsRequest{
+			Selector: &pb.ResourceSelection{
+				Resource: &pb.Resource{
+					Namespace: "testnamespace",
+				},
+			},
+		})
+
+		err = verifyPromQueries(&mockProm, "testnamespace")
+
+		if err != nil {
+			t.Fatalf("Prometheus queries don't match. Got error: %s", err)
+		}
+	})
+
+	t.Run("Reads namespace from resource name when resource is namespace", func(t *testing.T) {
+		k8sAPI, err := k8s.NewFakeAPI("", make([]string, 0)...)
+		if err != nil {
+			t.Fatalf("NewFakeAPI returned an error: %s", err)
+		}
+
+		mockProm := MockProm{Res: model.Vector{
+			&model.Sample{
+				Metric:    model.Metric{"pod": "emojivoto-meshed"},
+				Timestamp: 456,
+			},
+		}}
+		fakeGrpcServer := newGrpcServer(
+			&mockProm,
+			tap.NewTapClient(nil),
+			k8sAPI,
+			"linkerd",
+			[]string{},
+		)
+
+		k8sAPI.Sync()
+
+		_, _ = fakeGrpcServer.ListPods(context.TODO(), &pb.ListPodsRequest{
+			Selector: &pb.ResourceSelection{
+				Resource: &pb.Resource{
```

Comment:
More of a nit than anything else: I think it would be good to also set the `Namespace` field here to something other than `testnamespace` to further prove that the `Name` field is used instead of `Namespace` when `Type` is Namespace.

Reply:
@dadjeibaah this is actually very ambiguous :) Now that you're pointing it out... I think that having both namespace and type: Namespace set in the same request should be probably prohibited, right? Because it's not really obvious which one would be used...

Reply:
That's right. I think you actually have done that in the test above called `Fails for both resource and namespace set in request` now that I look at this more So we may not need this. I wasn't looking hard enough :man_facepalming: